jax/interpreters/pxla.py

```diff
@@ -123,31 +134,46 @@ def spec_to_indices(shape: Tuple[int, ...],
 
   Returns:
     A tuple of length `prod(sharding_spec.shards_per_axis) *
-    sharding_spec.replication_factor`.  Each element is an int, a slice object
-    with step=1, or a tuple thereof, to be treated as an index into the full
-    logical array.
+    prod(factor for factor, index in sharding_spec.replication_factors)`. Each
+    element is an int, a slice object with step=1, or a tuple thereof, to be
+    treated as an index into the full logical array.
   """
   assert len(shape) == len(sharding_spec.shards_per_axis)
-  indices_per_axis = [
-      _axis_indices(axis_size, num_shards, is_materialized)
-      for axis_size, num_shards, is_materialized in zip(
-          shape, sharding_spec.shards_per_axis, sharding_spec.is_axis_materialized)]
-
-  # Remove trailing slice(None) indices. This simplifies the final indices and
-  # hopefully makes them more likely to match what's passed in by the user in
-  # ShardedDeviceArray.__getitem__.
-  while len(indices_per_axis) > 1 and indices_per_axis[-1] == [slice(None)]:
-    indices_per_axis.pop(-1)
-
-  # `product` will always return a sequence of tuples. Skip the tuples if each
-  # index is a single element.
-  if len(indices_per_axis) == 1:
-    indices = list(indices_per_axis[0])
-  else:
-    indices = list(product(*indices_per_axis))
-
-  return tuple(i for i in indices
-               for _ in range(sharding_spec.replication_factor))
+  if not shape:
+    total_replication_factor = int(prod(factor for factor, index in
+                                        sharding_spec.replication_factors))
+    return ((),) * total_replication_factor
+  replication_factors = sorted(sharding_spec.replication_factors,
+                               key=op.itemgetter(1))
+
+  logical_index = 0
+  indices_per_mesh_axis = []
+  for mesh_index in range(len(shape) + len(sharding_spec.replication_factors)):
+    if replication_factors and replication_factors[0][1] == logical_index:
+      factor, _ = replication_factors.pop(0)
+      indices_per_mesh_axis.append([None] * factor)
```

Comment:
Maybe add a comment like "Insert placeholder `None` indices. We will remove these later, as they don't correspond to a logical index."